src/backend/distributed/master/master_cache_table_shards.c

```diff
@@ -0,0 +1,191 @@
+#include "distributed/metadata_cache.h"
+#include "distributed/multi_join_order.h"
+#include "distributed/pg_dist_shard.h"
+#include "distributed/worker_manager.h"
+#include "distributed/worker_protocol.h"
+#include "utils/builtins.h"
+#include "utils/lsyscache.h"
+
+
+static List * FindAbsentShardPlacementsOnWorker(WorkerNode *workerNode,
+												ShardInterval **shardIntervalArray,
+												List **placementListArray,
+												int shardCount);
+static void DropShardsFromWorker(WorkerNode *workerNode, Oid relationId,
+								 List *shardIntervalList);
+
+PG_FUNCTION_INFO_V1(master_expire_table_cache);
+
+
+/*
+ * master_expire_table_cache drops table's caches shards in all workers. The function
+ * expects a passed table to be a small distributed table meaning it has less than
+ * large_table_shard_count.
+ */
+Datum
+master_expire_table_cache(PG_FUNCTION_ARGS)
+{
+	Oid relationId = PG_GETARG_OID(0);
+	DistTableCacheEntry *cacheEntry = DistributedTableCacheEntry(relationId);
+	List *workerNodeList = WorkerNodeList();
+	ListCell *workerNodeCell = NULL;
+	int shardCount = cacheEntry->shardIntervalArrayLength;
+	ShardInterval **shardIntervalArray = cacheEntry->sortedShardIntervalArray;
+	List **placementListArray = NULL;
+	int shardIndex = 0;
+
+	if (shardCount == 0)
+	{
+		ereport(WARNING, (errmsg("Table has no shards, no action is taken")));
+		PG_RETURN_VOID();
+	}
+
+	if (shardCount >= LargeTableShardCount)
+	{
+		ereport(ERROR, (errmsg("Must be called on tables smaller than %d shards",
+							   LargeTableShardCount)));
+	}
+
+	placementListArray = palloc(shardCount * sizeof(List *));
+
+	for (shardIndex = 0; shardIndex < shardCount; shardIndex++)
+	{
+		ShardInterval *shardInterval = shardIntervalArray[shardIndex];
+		placementListArray[shardIndex] =
+			FinalizedShardPlacementList(shardInterval->shardId);
+	}
+
+	foreach(workerNodeCell, workerNodeList)
+	{
+		WorkerNode *workerNode = (WorkerNode *) lfirst(workerNodeCell);
+		List *shardDropList = FindAbsentShardPlacementsOnWorker(workerNode,
+																shardIntervalArray,
+																placementListArray,
+																shardCount);
+		DropShardsFromWorker(workerNode, relationId, shardDropList);
+	}
+
+	pfree(placementListArray);
+
+	PG_RETURN_VOID();
+}
+
+
+/*
+ * FindAbsentShardPlacementsOnWorker compiles shard interval list of shards
+ * that do not have registered placement at given worker node.
+ */
+List *
+FindAbsentShardPlacementsOnWorker(WorkerNode *workerNode,
+								  ShardInterval **shardIntervalArray,
+								  List **placementListArray, int shardCount)
+{
+	List *absentShardIntervalList = NIL;
+
+	int shardIndex = 0;
+	for (shardIndex = 0; shardIndex < shardCount; shardIndex++)
+	{
+		ShardInterval *shardInterval = shardIntervalArray[shardIndex];
+		List *placementList = placementListArray[shardIndex];
+
+		ListCell *placementCell = NULL;
+		foreach(placementCell, placementList)
+		{
+			ShardPlacement *shardPlacement = (ShardPlacement *) lfirst(
```

Comment:
This looks kind of funky. Can we avoid the line break if we use `placement` as the variable name? Or maybe break the assignment up across two lines, like…

```c
ShardPlacement *shardPlacement = NULL;
shardPlacement = (ShardPlacement *) lfirst(placementCell);
```